make sure to update the (core) proposals that are already created

MaximilianMeister opened the pull request:
cc @rhafer

vuntz commented:
+1

rhafer commented:
Somehow I think it would be much better to restore the core proposals before actually running the installer, if that is somehow possible. Otherwise it will be every easy to break the upgrade in weird ways. Especially if the network barclamp settings  before running the upgrade don't match exactly what's in the backup (post install changes of that are not really working).

E.g. in (simplified) steps:
1. Restore core baclamps proposals to /etc/crowbar/ and run the migrations on the json files. I guess we just need to put the "attributes" section into /etc/crowbar/
2. Run installer
3. Restore non core barclamps and update the "deployment" sections of the core barclamps 

@vuntz I'd really appreciate your thought here as well.

rhafer commented on crowbar_framework/lib/crowbar/backup/restore.rb:
```diff
@@ -47,7 +47,9 @@ def restore_chef
                 bc_name, prop = filename.split("-")
                 prop.gsub!(/.json$/, "")
-                Proposal.create(barclamp: bc_name, name: prop, properties: record.raw_data)
+                proposal = Proposal.first_or_initialize(barclamp: bc_name, name: prop)
```
Somehow this doesn't seem to work. While this now updates existing proposals successfully it somehow does not create new one if there wasn't one before. Didn't figure out why this happens yet.

rhafer replied:
It needs to be:

```Proposal.where(barclamp: bc_name).first_or_initialize(barclamp: bc_name, name: prop)```

Otherwise we will just overwrite the first Proposal (which ever that is) again and again.

MaximilianMeister replied:
thanks I changed it now to
```ruby
proposal = Proposal.where(barclamp: bc_name, name: prop).first_or_initialize
```

vuntz commented:
@rhafer hrm, I think you're correct, especially for the network barclamp.

It however means changes to the install process because install-suse-cloud will try to enforce network.json, I think...

rhafer commented:
@vuntz Doesn't https://github.com/crowbar/crowbar/blob/master/scripts/install-chef-suse.sh#L821 mean that at least crowbar, dns, network, provisioner and ntp are also supported?

vuntz commented:
@rhafer yes, but from all of these, only network has settings which cannot be changed after the install. All the other ones should work fine with the current workflow (although I agree it's suboptimal).

vuntz commented on crowbar_framework/lib/crowbar/upgrade.rb:
```diff
@@ -52,6 +54,22 @@ def knife_files
+              "echo",
+              content.read,
+              ">",
+              "/etc/crowbar/network.json"
```
Why not copy the file? Sounds a bit safer than redirecting the output of an `echo`.

MaximilianMeister replied:
you're right. will do

rhafer commented on crowbar_framework/lib/crowbar/backup/restore.rb:
```diff
@@ -47,7 +47,9 @@ def restore_chef
                 bc_name, prop = filename.split("-")
                 prop.gsub!(/.json$/, "")
-                Proposal.create(barclamp: bc_name, name: prop, properties: record.raw_data)
+                proposal = Proposal.where(barclamp: bc_name, name: prop).first_or_initialize
```
I think this is still wrong. See e.g.:

```https://github.com/crowbar/crowbar-core/blob/master/crowbar_framework/app/models/service_object.rb#L775``` for how it is called elsewhere.

MaximilianMeister replied:
right, sorry next time i will test it before. I thought Rails would be so smart to know what I want... will fix

MaximilianMeister commented on crowbar_framework/lib/schema_migration.rb:
```diff
@@ -58,6 +58,28 @@ def self.run_for_bc bc_name
+    end
+
+    proposal = Proposal.where(barclamp: bc_name, name: prop).first
+    proposal.raw_data = json
```
@rhafer can we just feed the databag here into raw_data?

rhafer replied:
Doesn't just work if you pass ```json``` instead of ```proposal``` into ```migrate_proposal```

jsuchome commented on crowbar_framework/lib/crowbar/upgrade.rb:
```diff
@@ -84,5 +104,11 @@ def filecontent_replace(file, search, replace)
+
+    def validate_proposal(barclamp, proposal)
+      so = ServiceObject.new(Rails.logger)
+      so.bc_name = barclamp
```
Is this required? Shouldn't it be created by ServiceObject?

MaximilianMeister replied:
default is "unknown"

jsuchome replied:
Ah, OK

rhafer commented:
Sorry, but can we leave the validation part out for now and handle that in a separate PR? Sorry for even bringing it up here. It requires some more thoughts and we should really get the rest of this merged.

rhafer commented on crowbar_framework/lib/schema_migration.rb:
```diff
@@ -58,6 +58,23 @@ def self.run_for_bc bc_name
+    all_scripts = find_scripts_for_bc(bc_name)
+    return if all_scripts.empty?
+
+    attributes = json["attributes"]
```
This needs to be ```json["attributes"][bc_name]```

rhafer commented on crowbar_framework/lib/schema_migration.rb:
```diff
@@ -58,6 +58,23 @@ def self.run_for_bc bc_name
+    return if all_scripts.empty?
+
+    attributes = json["attributes"]
+    deployment = json["deployment"]
```
Same here: json["deployment"][bc_name]

rhafer commented on crowbar_framework/lib/crowbar/upgrade.rb:
```diff
@@ -52,6 +54,24 @@ def knife_files
+        if file.basename.to_s =~ /^bc-network-(.*)\.json$/
+          json = JSON.load(file.read)
+          # override with migrated attributes
+          json["attributes"] = SchemaMigration.migrated_attributes("network", json)
```
With the suggested changes in ```migrated_attributes``` you need to assign the result to json["attributes"]["network"].

vuntz commented on crowbar_framework/lib/crowbar/upgrade.rb:
```diff
@@ -52,6 +54,22 @@ def knife_files
+          # override with migrated attributes
+          json["attributes"] = SchemaMigration.migrated_attributes("network", json)
+          Tempfile.open("network") do |content|
+            content.write JSON.pretty_generate(json["attributes"])
```
I was confused for one second because `content` is actually not the content, but a file object. Would it make sense to rename that variable?

rhafer replied:
You probably just need to reassigne the result of migrated_attributes to a new Hash.

```
attributes_migrated = {}
attributes_migrated["attributes"] = SchemaMigration.migrated_attributes("network", json)
```
 and then ```JSON.pretty_generate(attributes_migrated)```

I haven't tried that though.

rhafer commented on crowbar_framework/lib/crowbar/upgrade.rb:
```diff
@@ -52,6 +54,24 @@ def knife_files
+          # override with migrated attributes
+          json["attributes"] = SchemaMigration.migrated_attributes("network", json)
+          Tempfile.open("network") do |content|
+            content.write JSON.pretty_generate(json["attributes"])
```
We need the written json file to start with 
```
{
  "attributes":
    "network":
....
}
```

Currently it starts at ``` { "network": ... } ```

rhafer commented on crowbar_framework/lib/crowbar/upgrade.rb:
```diff
@@ -52,6 +54,24 @@ def knife_files
+            system(*cmd)
+          end
+          # finally validate network proposal
+          validate_proposal("network", json)
```
As said, lets leave this out for now and worry about that later.

MaximilianMeister commented:
> Sorry, but can we leave the validation part out for now and handle that in a separate PR? Sorry for even bringing it up here. It requires some more thoughts and we should really get the rest of this merged.

sure, i took out the commit.

I also addressed all the comments

rhafer commented on crowbar_framework/lib/crowbar/upgrade.rb:
```diff
@@ -52,6 +54,22 @@ def knife_files
+              f.path,
+              "/etc/crowbar/network.json"
+            ]
+            system(*cmd)
```
For some reason the copy in /etc/crowbar/ is of size 0 in my test setup. I gues you're missing the sync or close or something like that before actually copying the file here.

rhafer replied:
Confirmed. If I put ```f.close``` before the cmd the file in not empty anymore. But I am sure there's a more correct way to address this.

vuntz commented:
+1

vuntz commented on crowbar_framework/lib/schema_migration.rb:
```diff
@@ -58,6 +58,23 @@ def self.run_for_bc bc_name
+    deployment = json["deployment"][bc_name]
+
+    # return attributes
+    migrate_object(bc_name, template, all_scripts, attributes, deployment).first
```
Actually: `migrate_object` can return `nil`. Should we deal with this?

MaximilianMeister replied:
probably we should, would this be fatal then? wondering what do do in that case

MaximilianMeister replied:
this is still the wrong place for using `.first` will fix

rhafer commented:
+1

MaximilianMeister commented:
@rhafer before merging let me propose another commit on top

MaximilianMeister commented:
> Should this be localizable? (Also, I think it's better to put "Cannot" than "Could not" in error messages)

when the i18n namespaces for the controllers are done we can add this into the step that it belongs to I would say.

rhafer commented on crowbar_framework/lib/crowbar/upgrade.rb:
```diff
@@ -57,6 +63,27 @@ def knife_files
+        when /^bc-network-(.*)\.json$/
+          json = JSON.load(file.read)
+          # override with migrated attributes
+          attributes = SchemaMigration.migrate_proposal_from_json("network", json)
```
I guess you should rename ```attributes``` here. ```migrate_proposal_from_json``` does no longer return only the attributes.

MaximilianMeister replied:
done and rebased

rhafer commented:
+1

rhafer commented:
@vuntz are you still good with this?

vuntz commented:
yes, still +1.
